### PR TITLE
Add more descriptive error message to sys.save

### LINKED ISSUE #6532: Include error information when a file operation fails
**Is your feature request related to a problem? Please describe (REQUIRED):**
When sys.save() fails it would be nice to get more information about why it fails and not only "Could not open the file foo.txt".

**Describe the solution you'd like (REQUIRED):**
Perhaps add `errno` to the error message?

### PR DESCRIPTION
**Pull request by Jhonnyg**
If a `sys.save()` operation fails the error message will now contain additional information about the reason.

This PR adds a new function in `dstrings.h` (`dmStrError`) that translates and copies errno values to a buffer for printing human readable error messages.

Fixes #6532

### CODE REVIEW

File: engine/script/src/script_sys.cpp [hunk 1]
```diff
@@ -119,7 +119,7 @@ union SaveLoadBuffer
         FILE* file = fopen(tmp_filename, "wb");
         if (!file)
         {
-            return luaL_error(L, "Could not open the file %s.", tmp_filename);
+            return luaL_error(L, "Could not open the file %s, reason: %s.", tmp_filename, strerror(errno));
```
Comment (Jhonnyg) [inline engine/script/src/script_sys.cpp:122]: strerror is not thread safe, but we use it http_cache.cpp. an alternative would be strerror_r for posix systems, but I don't know how portable it is?

Comment (britzl) [reply engine/script/src/script_sys.cpp:122]: It seems like strerror_r has been around for quite some time:

https://man7.org/linux/man-pages/man3/strerror_l.3.html

Perhaps better to use it?

Comment (JCash) [reply engine/script/src/script_sys.cpp:122]: The strerror_r is not standard, and might not be available on all platforms (MSVC, Switch, PS4/PS5?)

A newer function is called `strerror_s`, available from `C11`.
I'm not sure what the current supported set of features from our compilers is.
We'd have to test it.

Another option is to grab the source from somewhere, and put it in our code (license permitting).

Comment (Jhonnyg) [reply engine/script/src/script_sys.cpp]: Can't use strerror_s on most platforms, but it seems I can use strerror_r in various forms depending on platform..

File: engine/dlib/src/dlib/dstrings.cpp [hunk 2]
```diff
@@ -231,3 +232,97 @@ int dmStrCaseCmp(const char *s1, const char *s2)
...
+        else
+        {
+            dmSnPrintf(scratch, scratch_size, "Failed getting error (code %d)", ret);
+        }
+    }
+
+    retstr = &scratch[0];
+#endif
+
+#ifdef DM_STRERROR_USE_GNU
```
Comment (JCash) [inline engine/dlib/src/dlib/dstrings.cpp:302]: Perhaps these could use "#elif defined(...", in order to ensure they aren't accidentally enabled at the same time?

File: engine/dlib/src/dmsdk/dlib/dstrings.h [hunk 3]
```diff
@@ -132,4 +132,23 @@ size_t dmStrlCat(char *dst, const char *src, size_t size);
...
+ * @param size Max size of destination string in bytes
+ * @return a null-terminated error message
+ * @examples
+ *
+ * ```cpp
+ * char buf[128];
+ * dmStrerror(buf, sizeof(buf), ENOENT); // buf => "No such file or directory"
+ * ```
+ */
+void dmStrerror(char* dst, size_t size, int err);
```
Comment (JCash) [inline engine/dlib/src/dmsdk/dlib/dstrings.h:152]: For consistency in this file, perhaps this could be named `dmStrError` ?

**Review by JCash — CHANGES_REQUESTED** [review]
I suggest a small rename of the function.

**Pull request merged**